boot/freeldr/freeldr/arch/drivers/hwide.c

```diff
@@ -2,940 +2,1252 @@
+        SectorCount -= BlockCount;
 
-    /* Allocate a new device unit structure */
-    *DeviceUnit = FrLdrTempAlloc(sizeof(DEVICE_UNIT), TAG_ATA_DEVICE);
-    if (*DeviceUnit == NULL)
-    {
-        ERR("Failed to allocate device unit!\n");
-        return FALSE;
+        Buffer = (PVOID)((ULONG_PTR)Buffer + Unit->P.SectorSize);
     }
 
-    RtlZeroMemory(*DeviceUnit, sizeof(DEVICE_UNIT));
-    (*DeviceUnit)->Channel = Channel;
-    (*DeviceUnit)->DeviceNumber = DeviceNumber;
-    (*DeviceUnit)->IdentifyData = Id;
+    return TRUE;
+}
 
-    if (Flags & ATA_DEVICE_ATAPI)
-    {
-        /* Clear the ATAPI 'Bus reset' indication */
-        for (i = 0; i < 10; ++i)
-        {
-            AtapiRequestSense(*DeviceUnit, &SenseData);
-            StallExecutionProcessor(10);
-        }
+PDEVICE_UNIT
+AtaGetDevice(
+    _In_ UCHAR UnitNumber)
+{
+    if (UnitNumber < RTL_NUMBER_OF(AtapUnits))
+        return (PDEVICE_UNIT)AtapUnits[UnitNumber];
 
-        /* Detect a medium's capacity */
-        AtapiCapacityDetect(*DeviceUnit, &TotalSectors, &SectorSize);
-        if (SectorSize == 0 || TotalSectors == 0)
-        {
-            /* It's ok and can be used to show alert like "Please insert the CD" */
-            TRACE("No media found.\n");
-            Flags |= ATA_DEVICE_NO_MEDIA;
-        }
-    }
-    else
+    return NULL;
+}
+
+BOOLEAN
+AtaInit(
+    _Out_ PUCHAR DetectedCount)
+{
+    ULONG Channel;
+
+    *DetectedCount = 0;
+
+    /* Enumerate IDE channels */
+    for (Channel = 0; Channel < CHANNEL_MAX_CHANNELS; ++Channel)
     {
-        if (Id.SupportLba || (Id.MajorRevision && Id.UserAddressableSectors))
-        {
-            if (Id.FeaturesSupport.Address48)
-            {
-                TRACE("Using LBA48 addressing mode.\n");
-                Flags |= ATA_DEVICE_LBA48 | ATA_DEVICE_LBA;
-                TotalSectors = Id.UserAddressableSectors48;
-            }
-            else
-            {
-                TRACE("Using LBA28 addressing mode.\n");
-                Flags |= ATA_DEVICE_LBA;
-                TotalSectors = Id.UserAddressableSectors;
-            }
+        UCHAR DeviceNumber;
+        IDE_REGISTERS Registers;
 
-            /* LBA ATA drives always have a sector size of 512 */
-            SectorSize = 512;
-        }
-        else
+        if (!AtapIdentifyChannel(Channel, &Registers))
+            continue;
+
+        /* Check for devices attached to the bus */
+        for (DeviceNumber = 0; DeviceNumber < CHANNEL_MAX_DEVICES; ++DeviceNumber)
         {
-            TRACE("Using CHS addressing mode.\n");
-            Flags |= ATA_DEVICE_CHS;
+            PHW_DEVICE_UNIT DeviceUnit;
+            ATA_DEVICE_CLASS DeviceClass;
 
-            if (Id.UnformattedBytesPerSector == 0)
-            {
-                SectorSize = 512;
-            }
-            else
+            /* Allocate a new device unit structure */
+            DeviceUnit = FrLdrTempAlloc(sizeof(*DeviceUnit), TAG_ATA_DEVICE);
```

Comment:
Question: The DeviceUnits are heap-allocated, like in the previous version of the code.
Did you decide  to not have a AtaFree() function (that would manually free these) because, these structures will automatically be freed when the freeldr _temp_ heap gets cleared up just before ReactOS gets booted?

Reply:
makes sense.